### ISSUE TITLE
[DI] [DX] Default service tag configuration causes confusion

### ISSUE THREAD

**Issue by curry684** [symfony/symfony#22497]
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | possibly
| BC Break report? | no
| RFC?             | yes
| Symfony version  | master

I just created a service definition file like this SCCEE:
```yaml
services:
    _defaults:
        tags:           [ { name: kernel.event_subscriber } ]

    some.event.subscriber:
        calls:
            - [setLogger, ["@logger"]]
        tags:           [ { name: monolog.logger, channel: specific_channel } ]
```
And then spent half an hour debugging why my events weren't being processed.

The reasoning is sound of course - I provided a default, and then I provided a real value overwriting the default. It's however not intuitive and imho highly confusing that this doesn't work like this out of the box.

I see 2 solutions to this:
1. Merge tags by default based on name. Seems to make sense but it may go wrong in edge cases I'm not seeing.
2. If a 'mergeable' property like `tags` is provided both in default/instanceof and definition, require setting a property in the like `merge_tags` explicitly to true or false, and bail out if it isn't set.

Comments?

**Comment by nicolas-grekas** [comment]
If you want to get tags from defaults while defining more on a definition, you need to add I herit_tags: true.
The current behavior is on purpose and consistent with how child/parent defs do not inherit tags by default.
There is no one fits all merging strategy, that's why things are how they are now.
Doc PR welcomed to make easier to understand.

**Comment by curry684** [comment]
Hrm I actually tried `inherit_tags: true` and found it not to be working, as I had expected as technically `_defaults` is more like a PHP trait to me and not true inheritance - there's no parent to inherit from.

However I tried to add it to the `_defaults` section instead of the service itself, where it indeed is working, but that kind of defeats the point as that means I'm still bloating my config with an extra property on every service.

How about the following:
1. We should support `inherit_tags` in the `_defaults` and `_instanceof` sections as well to reduce the boilerplate.
2. We should issue a warning during container compilation if multiple tag sets are applicable to a service and the behavior is not explicitly defined.

The warning could be something like:
> Service 'some.event.subscriber' defines tags, which overwrite tags defined in applicable _defaults or _instanceof sections. Specify the inherit_tags property explicitly to state your intended result and remove this warning.

**Comment by nicolas-grekas** [comment]
> add it to the _defaults section instead of the service itself

that's what I meant sorry if that wasn't explicit.

About the rest, we're just talking about configuration. We added helpers to make config simpler in 3.3, but by no mean we're trying to turn that into a fully expressive DSL. Reducing the boilerplate for the shake of it is not a target. If the shortcuts don't fit your need, just use traditional wiring...

About multiple tags, it's a compiler-pass-per-compiler-pass decision - I don't see what we could do at a lower and more generic level.

**Comment by nicolas-grekas** [comment]
(should we just remove "tags" from "_defaults"?)

**Comment by curry684** [comment]
I've forward-ported a big project to 3.3 as that's going to be long released before the project itself. And I just took the blogs, read through all the new features, and started using them where it would clarify my code or ease my maintenance. So I completely emptied my `services.yml` files, replaced them with a simple `imports: { services/* }` rule, and started building grouped service blocks in that folder. My `voters.yml` looks like this:
```yaml
services:
    _defaults:
        public: false
        tags: [ security.voter ]

    MyBundle\Security\Voter\PostVoter: ~
    MyBundle\Security\Voter\UserVoter: ~
    ...etc
```
I could have also done this with `_instanceof` in a single file, that's just personal preference on my part. I'm just using `_defaults` as an `_instanceof.*` file-based convenience rule.

I went into this assuming that *all* properties would be defaultable to begin with, and was highly surprised it wasn't the case. I see no real reason not to allow `arguments` and `calls` as well. I've had more DX issues so far based on, to me arbitrarily chosen, properties *not* being defaulted than the other way around. I also tried this at some point:
```yaml
services:
  _instanceof:
    Psr\Log\LoggerAwareInterface:
      calls: [ [ setLogger, [ "@logger" ] ] ]
```
Which doesn't work as `calls` isn't allowed. I've encountered similar issues with `factory` for my Doctrine repositories as discussed elsewhere. So to me from a Symfony-consumer perspective: I've been given this great new box of toys to play with, and then found out along the way half the toys in the box were intentionally broken. What's left if we also remove `tags` from the box? We'll be left with this great new feature which more or less supports defaulting..... `autowire`. Which most of us don't use in production code as it is incredibly magic by its very definition.

I think the right way is not limiting the feature (even more), but opening it wide up. Allow defaulting all properties, allow all properties in `_instanceof`. People *will* find a use case for default `arguments`, just like we did for `factory` and my `calls` example above. Symfony is about enabling, not limiting. We'll just need to think properly on how to resolve the array based properties.
